### ISSUE TITLE
Dataset#set_segment_size() hangs

### ISSUE THREAD

**Issue by sheaconlon** [sheaconlon/cell_counting#8]
Dataset#set_segment_size() sometimes hangs, writing out segments indefinitely. This seems to happen when the number of examples is a multiple of the segment size.

**Comment by sheaconlon** [comment]
I've produced an example. This writes data to disk at about 10MB/s indefinitely. If `7` is used as the parameter to `set_segment_size`, then the program finishes in a second or two.

```python
import sys
sys.path.append("/Users/sheaconlon/Dropbox/research/cell_counting")

from src import dataset

images_masks = dataset.Dataset(1)
images_masks.load_image_mask_pairs(
	"/Users/sheaconlon/Dropbox/research/cell_counting/data/easy/raw/images",
    "/Users/sheaconlon/Dropbox/research/cell_counting/data/easy/raw/masks",
    (2448*2, 2448*2))

def map_fn(example):
	return ((i, i + 1) for i in range(1, 10 + 1))
images_masks.map(map_fn)
images_masks.set_segment_size(10)
```

**Comment by sheaconlon** [comment]
After SIGINT, this is the traceback.

```
^CTraceback (most recent call last):
  File "mcve.py", line 15, in <module>
    images_masks.set_segment_size(10)
  File "/Users/sheaconlon/Dropbox/research/cell_counting/src/dataset.py", line 298, in set_segment_size
    new_dataset._add_segment(inputs, outputs)
  File "/Users/sheaconlon/Dropbox/research/cell_counting/src/dataset.py", line 382, in _add_segment
    np.save(os.path.join(this_segment_dir, "inputs.npy"), inputs)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/numpy/lib/npyio.py", line 511, in save
    pickle_kwargs=pickle_kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/numpy/lib/format.py", line 594, in write_array
    array.tofile(fp)
KeyboardInterrupt
```

**Comment by sheaconlon** [comment]
Resolved by bf3872c2093f8ff49a6d0c7197a96b733ef9228d.
